packages/battery_plus/tizen/src/battery_plus_tizen_plugin.cc

```diff
@@ -4,199 +4,131 @@
-    plugin->SetupChannels(registrar);
-    registrar->AddPlugin(std::move(plugin));
+#include "device_battery.h"
+
+namespace {
+
+typedef flutter::EventChannel<flutter::EncodableValue> FlEventChannel;
+typedef flutter::EventSink<flutter::EncodableValue> FlEventSink;
+typedef flutter::MethodCall<flutter::EncodableValue> FlMethodCall;
+typedef flutter::MethodResult<flutter::EncodableValue> FlMethodResult;
+typedef flutter::MethodChannel<flutter::EncodableValue> FlMethodChannel;
+typedef flutter::StreamHandler<flutter::EncodableValue> FlStreamHandler;
+typedef flutter::StreamHandlerError<flutter::EncodableValue>
+    FlStreamHandlerError;
+
+std::string BatteryStatusToString(BatteryStatus status) {
+  switch (status) {
+    case BatteryStatus::kCharging:
+      return "charging";
+    case BatteryStatus::kFull:
+      return "full";
+    case BatteryStatus::kDischarging:
+      return "discharging";
+    case BatteryStatus::kUnknown:
+    default:
+      return "unknown";
   }
+}
 
-  BatteryPlusTizenPlugin() {}
-
-  virtual ~BatteryPlusTizenPlugin() {}
-
-  void RegisterObserver(
-      std::unique_ptr<flutter::EventSink<flutter::EncodableValue>> &&events) {
+class BatteryStatusStreamHandler : public FlStreamHandler {
+ protected:
+  std::unique_ptr<FlStreamHandlerError> OnListenInternal(
+      const flutter::EncodableValue *arguments,
+      std::unique_ptr<FlEventSink> &&events) override {
     events_ = std::move(events);
 
-    // DEVICE_CALLBACK_BATTERY_CHARGING callback is called in only two cases
-    // like charging and discharing. When the charging status becomes "Full",
-    // discharging status event is called. So if it is full and disconnected
-    // from USB or AC charger, then any callbacks will not be called because the
-    // status already is the discharging status. To resolve this issue,
-    // DEVICE_CALLBACK_BATTERY_LEVEL callback is added. This callback can check
-    // whether it is disconnected in "Full" status. That is, when the battery
-    // status is full and disconnected, the level status will be changed from
-    // DEVICE_BATTERY_LEVEL_FULL to DEVICE_BATTERY_LEVEL_HIGH.
-    int ret = device_add_callback(DEVICE_CALLBACK_BATTERY_CHARGING,
-                                  BatteryChangedCB, this);
-    if (ret != DEVICE_ERROR_NONE) {
-      events_->Error("failed_to_add_callback", get_error_message(ret));
-      return;
+    BatteryStatusCallback callback = [this](BatteryStatus status) -> void {
+      if (status != BatteryStatus::kError) {
+        events_->Success(
+            flutter::EncodableValue(BatteryStatusToString(status)));
+      } else {
+        events_->Error(std::to_string(battery_.GetLastError()),
+                       battery_.GetLastErrorString());
+      }
+    };
+    if (!battery_.StartListen(callback)) {
+      return std::make_unique<FlStreamHandlerError>(
+          std::to_string(battery_.GetLastError()),
+          battery_.GetLastErrorString(), nullptr);
```

Comment:
This currently results in an app crash. There will be a fix in the engine code soon.

Issue: https://github.com/flutter/flutter/issues/101682

Reply:
@swift-kim Any progress on this?

Reply:
@bbrto21 It isn't likely the fix will land very soon. How do you think of [this](https://github.com/flutter-tizen/plugins/pull/387#discussion_r883346794) or [this](https://github.com/flutter-tizen/plugins/pull/387/commits/50086e595cd1839ae000f05c46868a4eacc3faba) workaround?

Reply:
@swift-kim Actually, I also made a patch to prevent crash. how about [this](https://github.com/bbrto21/plugins/commit/a351f1143a797ac3c5644e5b46d9c25d614df237)?
First one has intentional leak and the second solution you mentioned doesn't seem very good as it depends on instances of other classes.

Reply:
The `holder_` suffix seems redundant. Otherwise looks good to me.

Reply:
@swift-kim Can I make PR to apply all place where have same problem?

Reply:
Sure. Just one thing I forgot to mention: the `private:` section header is missing in the class.